### PR TITLE
Feature : add LOV Migration Script

### PR DESCRIPTION
**Pull request by muhammedBkf**

### CODE REVIEW

File: bin/lov_migrator [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+          name: graph.query([agent_uri, RDF::URI("http://xmlns.com/foaf/0.1/name"), nil]).first&.object,
+          sameAs: graph.query([agent_uri, RDF::URI("http://www.w3.org/2002/07/owl#sameAs"), nil]).map { |statement| extract_uri_or_value(statement.object) },
+        }
+      end
+
+    end
+
+    class Distribution
+
+      def extract_distribution_info(graph, distribution_uri)
```
Comment (syphax-bouazzouni) [inline bin/lov_migrator:125]: to update to work with the graph as an hash

File: bin/lov_migrator [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        info[:lastModifiedInLOVAt] = get_value(vocab, "http://purl.org/dc/terms/modified")
+
+        info
+      end
+
+    end
+
+    class Agent
+
+      def extract_agent_info(graph, agent_uri)
```
Comment (syphax-bouazzouni) [inline bin/lov_migrator:114]: to update to work with the  graph as an hash

File: bin/lov_migrator [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        end
+      else
+        value.to_s.empty? ? "None" : value.to_s
+      end
+    end
+
+  end
+end
+
+def print_vocabulary_info(info)
```
Comment (syphax-bouazzouni) [inline bin/lov_migrator:276]: to remove not needed in the final version

File: bin/lov_migrator [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+require 'rdf/n3'
+require 'csv'
+require 'open-uri'
+require 'digest'
+require 'benchmark'
+require 'parallel'
+
+CSV_MAIN_ATTRS = [ :prefix, :title, :description, :keyword, :creator, :uri, :lastModifiedInLOVAt ]
+CSV_ADDED_ATTRS = [ :destination, :who, :comment ]
+CSV_DISPATCH_FILENAME = 'LOV_vocabularies_dispatch.csv'
```
Comment (syphax-bouazzouni) [inline bin/lov_migrator:38]: this need to be an argument of the script

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]
I updated the code to be more structured into modules: `DumpParser`, `CSVGenerator` and a object for each of the models to migrate: `Vocabulary`, `Agent` and `Distribution`.

I optimized the n3 parsing code, to not save all the graph in memory, but just building a Hash, while streaming each line/statement on the graph.

Before optimization:
```
➜  ncbo_cron git:(lov_migrator) ✗ bin/lov_migrator --all 
File downloaded successfully as lov.n3.gz.tmp
The local file remains unchanged.
Parsing the n3 file in memory
Parsing the n3 file in memory ended in 150.80259921000106 seconds
Find vocabularies start
Find vocabularies ended in 0.02299319504527375 seconds
Found 861 vocabularies:
```

After optimization:
```
➜  ncbo_cron git:(lov_migrator) ✗ bin/lov_migrator --all 
File downloaded successfully as lov.n3.gz.tmp
The local file remains unchanged.
Parsing the n3 file in memory
Parsing the n3 file in memory ended in 11.830599968961906 seconds with 861 subjects found
Find vocabularies ended in 5.6899734772741795e-06 seconds 
```

Check my commits to see the details and understand the process.

**Comment by muhammedBkf** [comment]
> I updated the code to be more structured into modules: `DumpParser`, `CSVGenerator` and a object for each of the models to migrate: `Vocabulary`, `Agent` and `Distribution`.
> 
> I optimized the n3 parsing code, to not save all the graph in memory, but just building a Hash, while streaming each line/statement on the graph.
> 
> Before optimization:
> 
> ```
> ➜  ncbo_cron git:(lov_migrator) ✗ bin/lov_migrator --all 
> File downloaded successfully as lov.n3.gz.tmp
> The local file remains unchanged.
> Parsing the n3 file in memory
> Parsing the n3 file in memory ended in 150.80259921000106 seconds
> Find vocabularies start
> Find vocabularies ended in 0.02299319504527375 seconds
> Found 861 vocabularies:
> ```
> 
> After optimization:
> 
> ```
> ➜  ncbo_cron git:(lov_migrator) ✗ bin/lov_migrator --all 
> File downloaded successfully as lov.n3.gz.tmp
> The local file remains unchanged.
> Parsing the n3 file in memory
> Parsing the n3 file in memory ended in 11.830599968961906 seconds with 861 subjects found
> Find vocabularies ended in 5.6899734772741795e-06 seconds 
> ```
> 
> Check my commits to see the details and understand the process.
> 
> And add always a description to your PRs.

The time gain from using a ruby hash instead of loading the hash in memory is pretty impressive. I will update the script accordingly, thanks for the review.

**Comment by muhammedBkf** [comment]
It's closed, replaced by [PR#26](https://github.com/ontoportal-lirmm/ncbo_cron/pull/26)

**Pull request closed without merging**